include/envoy/thread/thread.h

```diff
@@ -2,40 +2,162 @@
 
 #include "envoy/common/pure.h"
 
+#include "absl/base/thread_annotations.h"
+
 namespace Envoy {
 namespace Thread {
 
 /**
- * Like the C++11 "basic lockable concept" but a concrete interface vs. a template.
+ * Like the C++11 "basic lockable concept" but a pure virtual interface vs. a template, and
+ * with thread annotations.
  */
-class BasicLockable {
+class LOCKABLE BasicLockable {
 public:
   virtual ~BasicLockable() {}
 
-  virtual void lock() PURE;
-  virtual bool try_lock() PURE;
-  virtual void unlock() PURE;
+  virtual void lock() EXCLUSIVE_LOCK_FUNCTION() PURE;
+  virtual bool tryLock() EXCLUSIVE_TRYLOCK_FUNCTION(true) PURE;
+  virtual void unlock() UNLOCK_FUNCTION() PURE;
 };
 
 /**
  * A lock guard that deals with an optional lock.
  */
-template <class T> class OptionalLockGuard {
+class SCOPED_LOCKABLE OptionalLockGuard {
+public:
+  /**
+   * Establishes a scoped mutex-lock. If non-null, the mutex is locked upon construction.
+   *
+   * @param lock the mutex.
+   */
+  OptionalLockGuard(BasicLockable* lock) EXCLUSIVE_LOCK_FUNCTION(lock) : lock_(lock) {
+    if (lock_ != nullptr) {
+      lock_->lock();
+    }
+  }
+
+  /**
+   * Destruction of the OptionalLockGuard unlocks the lock, if it is non-null.
+   */
+  ~OptionalLockGuard() UNLOCK_FUNCTION() {
+    if (lock_ != nullptr) {
+      lock_->unlock();
+    }
+  }
+
+private:
+  BasicLockable* const lock_;
+};
+
+// At the moment, TryLockGuard is very hard to annotate correctly, I
+// believe due to limitations in clang. At the moment there are no
+// GUARDED_BY variables for any tryLocks in the codebase, so it's
+// easies just to leave it out. In a future clang release it's
```

Comment:
typo "easies" (unless you are going for hip 😉)

Reply:
I wish I could take credit for that... done :)